File: Device/SFBSystemAudioObject.m
/*
 * Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
 * See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
 */

#import "SFBSystemAudioObject.h"
#import "SFBAudioObject+Internal.h"

@interface SFBSystemAudioObject ()
- (instancetype)init;
@end

@implementation SFBSystemAudioObject

static SFBSystemAudioObject *sSharedInstance = nil;

+ (SFBSystemAudioObject *)sharedInstance
{
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		sSharedInstance = [[SFBSystemAudioObject alloc] init];
	});
	return sSharedInstance;
}

- (instancetype)init
{
	if((self = [super init]))
		_objectID = kAudioObjectSystemObject;
	return self;
}

- (NSNumber *)mixStereoToMono
{
	return [self unsignedIntForProperty:kAudioHardwarePropertyMixStereoToMono inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (BOOL)setMixStereoToMono:(BOOL)value error:(NSError **)error
{
	return [self setUnsignedInt:(value != 0) forProperty:kAudioHardwarePropertyMixStereoToMono inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:error];
}

- (NSNumber *)processIsMaster
{
	return [self unsignedIntForProperty:kAudioHardwarePropertyProcessIsMaster inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSNumber *)isInitingOrExiting
{
	return [self unsignedIntForProperty:kAudioHardwarePropertyIsInitingOrExiting inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (BOOL)setUserIDChangedReturningError:(NSError **)error
{
	return [self setUnsignedInt:1 forProperty:kAudioHardwarePropertyUserIDChanged inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:error];
}

- (NSNumber *)processIsAudible
{
	return [self unsignedIntForProperty:kAudioHardwarePropertyProcessIsAudible inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (BOOL)setProcessIsAudible:(BOOL)value error:(NSError **)error
{
	return [self setUnsignedInt:(value != 0) forProperty:kAudioHardwarePropertyProcessIsAudible inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:error];
}

- (NSNumber *)sleepingIsAllowed
{
	return [self unsignedIntForProperty:kAudioHardwarePropertySleepingIsAllowed inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (BOOL)setSleepingIsAllowed:(BOOL)value error:(NSError **)error
{
	return [self setUnsignedInt:(value != 0) forProperty:kAudioHardwarePropertySleepingIsAllowed inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:error];
}

- (NSNumber *)unloadingIsAllowed
{
	return [self unsignedIntForProperty:kAudioHardwarePropertyUnloadingIsAllowed inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (BOOL)setUnloadingIsAllowed:(BOOL)value error:(NSError **)error
{
	return [self setUnsignedInt:(value != 0) forProperty:kAudioHardwarePropertyUnloadingIsAllowed inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:error];
}

- (NSNumber *)hogModeIsAllowed
{
	return [self unsignedIntForProperty:kAudioHardwarePropertyHogModeIsAllowed inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (BOOL)setHogModeIsAllowed:(BOOL)value error:(NSError **)error
{
	return [self setUnsignedInt:(value != 0) forProperty:kAudioHardwarePropertyHogModeIsAllowed inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:error];
}

- (NSNumber *)userSessionIsActiveOrHeadless
{
	return [self unsignedIntForProperty:kAudioHardwarePropertyUserSessionIsActiveOrHeadless inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (NSNumber *)powerHint
{
	return [self unsignedIntForProperty:kAudioHardwarePropertyPowerHint inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:NULL];
}

- (BOOL)setPowerHint:(AudioHardwarePowerHint)value error:(NSError **)error
{
	return [self setUnsignedInt:value forProperty:kAudioHardwarePropertyPowerHint inScope:kAudioObjectPropertyScopeGlobal onElement:kAudioObjectPropertyElementMaster error:error];
}

@end

File: Device/SFBSystemAudioObject.swift
/*
* Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
* See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
*/

import Foundation

extension SystemAudioObject {
	/// Returns `true` if audio devices should mix stereo to mono
	/// - note: This corresponds to `kAudioHardwarePropertyMixStereoToMono`
	/// - throws: An error if the property could not be retrieved
	public func mixStereoToMono() throws -> Bool {
		return try getProperty(.mixStereoToMono) != 0
	}

	/// Returns `true` if the current process contains the master HAL instance
	/// - note: This corresponds to `kAudioHardwarePropertyProcessIsMaster`
	/// - throws: An error if the property could not be retrieved
	public func processIsMaster() throws -> Bool {
		return try getProperty(.processIsMaster) != 0
	}

	/// Returns `true` if the HAL is initing or exiting the process
	/// - note: This corresponds to `kAudioHardwarePropertyIsInitingOrExiting`
	/// - throws: An error if the property could not be retrieved
	public func isInitingOrExiting() throws -> Bool {
		return try getProperty(.isInitingOrExiting) != 0
	}

	/// Returns `true` if the process will be heard
	/// - note: This corresponds to `kAudioHardwarePropertyProcessIsAudible`
	/// - throws: An error if the property could not be retrieved
	public func processIsAudible() throws -> Bool {
		return try getProperty(.processIsAudible) != 0
	}

	/// Returns `true` if the process will allow the CPU to sleep while audio IO is in progress
	/// - note: This corresponds to `kAudioHardwarePropertySleepingIsAllowed`
	/// - throws: An error if the property could not be retrieved
	public func sleepingIsAllowed() throws -> Bool {
		return try getProperty(.sleepingIsAllowed) != 0
	}

	/// Returns `true` if the process should be unloaded after a period of inactivity
	/// - note: This corresponds to `kAudioHardwarePropertyUnloadingIsAllowed`
	/// - throws: An error if the property could not be retrieved
	public func unloadingIsAllowed() throws -> Bool {
		return try getProperty(.unloadingIsAllowed) != 0
	}

	/// Returns `true` if the HAL should automatically take hog mode on behalf of the process
	/// - note: This corresponds to `kAudioHardwarePropertyHogModeIsAllowed`
	/// - throws: An error if the property could not be retrieved
	public func hogModeIsAllowed() throws -> Bool {
		return try getProperty(.hogModeIsAllowed) != 0
	}

	/// Returns `true` if the login session of the user is a console or headless session
	/// - note: This corresponds to `kAudioHardwarePropertyUserSessionIsActiveOrHeadless`
	/// - throws: An error if the property could not be retrieved
	public func userSessionIsActiveOrHeadless() throws -> Bool {
		return try getProperty(.userSessionIsActiveOrHeadless) != 0
	}

	/// Returns the power hint
	/// - note: This corresponds to `kAudioHardwarePropertyPowerHint`
	/// - throws: An error if the property could not be retrieved
	public func powerHint() throws -> AudioHardwarePowerHint {
		return AudioHardwarePowerHint(rawValue: try getProperty(.powerHint))!
	}
}
